### DISCUSSION TITLE
Empty black screen while trying to use the gui

### DISCUSSION THREAD

**Discussion by aRdA-dLy** [CosmosOS/Cosmos#2902] [Q&A]
#### Area of Cosmos - What area of Cosmos are we dealing with?
GUI

#### Expected Behaviour - What do you think that should happen?
A dark violet background with a gray rectangle at the bottom of the screen.

#### Actual Behaviour - What unexpectedly happens?
There isn't any color or shape in the screen. Just a black screen.

#### Reproduction - How did you get this error to appear?
I wrote the code and run it in VMware.

#### Version - Were you using the User Kit or Dev Kit? And what User Kit version or Dev Kit commit (Cosmos, IL2CPU, X#)?
User Kit
![Ek Açıklama 2024-01-13 232323](https://github.com/CosmosOS/Cosmos/assets/156545016/2f131022-4529-4f33-8f0a-085427fe916c)

**Comment by alessiob07** [comment]
Hello, can you write the code you used?
I think there is a missing Canvas.Display() on Run().

**Comment by aRdA-dLy** [comment]
**GUI.cs**

    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;
    using System.Threading.Tasks;
    using System.Drawing;
    using Cosmos.System.Graphics;
    using Cosmos.System;
    using Sys = Cosmos.System;
    
    namespace mondaOS.Graphics
    {
        public class GUI
        {
    
            private Canvas canvas;
            private Pen pen;
            private List<Tuple<Sys.Graphics.Point, Color>> savedPixels;
            private Tabbar tabbar;
    
            private MouseState prevMouseState;
    
            private UInt32 pX, pY;
    
            public GUI()
            {
                this.canvas = FullScreenCanvas.GetFullScreenCanvas();
                this.canvas.Clear(Color.DarkViolet);
    
                this.pen = new Pen(Color.White);
                this.prevMouseState = MouseState.None;
    
                this.pX = 3;
                this.pY = 3;
                this.savedPixels=new List<Tuple<Sys.Graphics.Point, Color>>();
    
                this.tabbar = new Tabbar(this.canvas);
    
                MouseManager.ScreenHeight = (UInt32)this.canvas.Mode.Rows;
                MouseManager.ScreenWidth = (UInt32)this.canvas.Mode.Columns;
            }
    
            public void handleGUIInputs()
            {
                if(this.pX!=MouseManager.X&&this.pY!=MouseManager.Y)
                {
                    if (MouseManager.X < 2 || MouseManager.Y < (MouseManager.ScreenWidth - 2) || MouseManager.Y > (MouseManager.ScreenHeight - 2))
                        return;
    
                    this.pX=MouseManager.X;
                    this.pY=MouseManager.Y;
    
                    Sys.Graphics.Point[] points = new Sys.Graphics.Point[]
                    {
                        new Sys.Graphics.Point((Int32)MouseManager.X,(Int32)MouseManager.Y),
                        new Sys.Graphics.Point((Int32)MouseManager.X+1,(Int32)MouseManager.Y),
                        new Sys.Graphics.Point((Int32)MouseManager.X-1,(Int32)MouseManager.Y),
                        new Sys.Graphics.Point((Int32)MouseManager.X,(Int32)MouseManager.Y+1),
                        new Sys.Graphics.Point((Int32)MouseManager.X,(Int32)MouseManager.Y-1)
    
    
                    };
                    
                    foreach (Tuple<Sys.Graphics.Point,Color> pixelData in this.savedPixels)
                        this.canvas.DrawPoint(new Pen(pixelData.Item2), pixelData.Item1);
    
                    this.savedPixels.Clear();
    
                    foreach (Sys.Graphics.Point p in points) {
    
                        this.savedPixels.Add(new Tuple<Sys.Graphics.Point, Color>(p, this.canvas.GetPointColor(p.X, p.Y)));
                        this.canvas.DrawPoint(this.pen, p);
    
    
                    }
                }
    
                if (MouseManager.MouseState == MouseState.Left&&this.prevMouseState!=MouseState.Left)
                    System.Console.Beep();
    
                this.prevMouseState = MouseManager.MouseState;
    
            }
    
        }
    }

**launchgui.cs**

      using System;
      using System.Collections.Generic;
      using System.Linq;
      using System.Text;
      using System.Threading.Tasks;
      using mondaOS.Graphics;
      
      namespace mondaOS.Commands
      {
          public class LaunchGUI : Command
          {
      
              public LaunchGUI (String name) : base (name) { }
      
              public override string execute(string[] args)
              {
                  if (Kernel.gui != null)
                      return "You are already in GUI.";
      
                  Kernel.gui = new GUI();
      
                  return "Launched GUI";
              }
      
      
          }
      }

**Kernel.cs**

      using System;
      using System.Collections.Generic;
      using System.Text;
      using Sys = Cosmos.System;
      using mondaOS.Commands;
      using Cosmos.System.FileSystem;
      using mondaOS.Graphics;
      
      namespace mondaOS
      {
          public class Kernel : Sys.Kernel
          {
              private CommandManager commandManager;
              private CosmosVFS vfs;
              public static GUI gui;
      
              protected override void BeforeRun()
              {
                  this.vfs = new CosmosVFS();
                  Sys.FileSystem.VFS.VFSManager.RegisterVFS(this.vfs);
                  this.commandManager = new CommandManager();
      
                  Console.WriteLine("Welcome mondaOS");
      
              }
      
              protected override void Run()
              {
                  if(Kernel.gui!= null)
                  {
                      Kernel.gui.handleGUIInputs();
                      return;
                  }
      
                  String response;
                  String input=Console.ReadLine();
                  response = this.commandManager.processInput(input);
                  Console.WriteLine(response);
              }
      
          }
      }

**Comment by alessiob07** [comment]
try add gui.canvas.Display(); at the end of Run() on Kernel.cs
It's needed to display to screen what is into graphics buffer.

**Comment by aRdA-dLy** [comment]
![image](https://github.com/CosmosOS/Cosmos/assets/156545016/1a8321a7-e0a8-4a5c-92f3-b63beaadae56)
There isn't any black mouse pointer or lines but the background and the taskbar is working

**Reply by 9xbt** [reply]
you have to add it after you render everything

**Reply by 9xbt** [reply]
so like
```cs
renderBackground();
renderTaskbar();
renderMouse();
canvas.Display();
```

**Reply by 9xbt** [reply]
oh and dont forget to render the mouse pointer last since its always on top of everything

**Reply by aRdA-dLy** [reply]
![image](https://github.com/CosmosOS/Cosmos/assets/156545016/eefdf2f7-3244-4ce7-a01a-9c396b30f52d)
How can I add them?

**Reply by blackneeed** [reply]
Add some methods

**Reply by aRdA-dLy** [reply]
What methods?

**Reply by blackneeed** [reply]
Actually you know what
Learn **C#**

**Reply by tails1154** [reply]
thats rude

**Comment by blackneeed** [comment]
This smells like from Tech Media tutorial

**Reply by GeorgeFence** [reply]
Tr

**Reply by blackneeed** [reply]
yeah
